bigflow/konfig.py

```diff
@@ -0,0 +1,280 @@
+
+    def __getitem__(self, k):
+        return getattr(self, k)
+
+    def __iter__(self):
+        return iter(vars(self).keys())
+
+    def __len__(self):
+        return len(vars(self))
+
+    # helpers & utils
+
+    @property
+    def name(self):
+        return type(self)._name
+
+    def replace(self, **kwargs):
+        """Return a new instance of the konfig replacing specified fields"""
+        cls = type(self)
+        return cls(**{**self._init_kwargs, **kwargs})
+
+    @classmethod
+    def resolve(
+        cls: tp.Type[K],
+        name: tp.Optional[str] = None,
+        default: tp.Optional[str] = None,
+        lazy: bool = True,
+        extra: tp.Optional[tp.Dict[str, tp.Any]] = None,
+    ) -> K:
+        return resolve_konfig(
+            [k for k in _all_subclasses(cls) if k._name],
+            name=name,
+            default=default,
+            lazy=lazy,
+            extra=extra,
+        )
+
+
+def _all_subclasses(cls: type):
+    yield cls
+    for sc in cls.__subclasses__():
+        yield from _all_subclasses(sc)
+
+
+class _LazyKonfig(lazy_object_proxy.Proxy):
+
+    def __init__(self, **kwargs):
+        super().__init__(lambda: resolve_konfig(**kwargs))
+
+    def __repr__(self):
+        return str(self)
+
+
+def resolve_konfig(
+    konfigs: tp.Union[
+        tp.Dict[str, tp.Type[K]],
+        tp.List[tp.Type[K]],
+    ],
+    name: tp.Optional[str] = None,
+    default: tp.Optional[str] = None,
+    lazy: bool = True,
+    extra: tp.Optional[tp.Dict[str, tp.Any]] = None,
+) -> K:
+
+    """Creates instance of config.
+
+    Konfigs may be passed as list (in such case Class.name is used as a konfig name)
+    or as a plain dictionary {name -> konfig_class}.
+
+    Konfig name may be provided explicitly with `name` argument or passed via `bigflow` cli tool.
+    Default konfig config name may be passed as a fallback.
+    """
+
+    if not isinstance(konfigs, tp.Dict):
+        for k in konfigs:
+            if not k._name:
+                raise ValueError(f"Konfig {k!r} does not have a name")
+        konfigs = {k._name: k for k in konfigs if k._name}
+    else:
+        konfigs = dict(konfigs)
+    extra = dict(extra or {})
+
+    if lazy:
+        return _LazyKonfig(konfigs=konfigs, name=name, default=default, extra=extra, lazy=False)
+
+    logger.debug("Resolve konfig %s from %s", name or "*", konfigs)
+    name = name or current_env() or default
+
+    if not name:
+        raise ValueError("Konfig name should not be empty")
+    logger.debug("Konfig name is %s", name)
+
+    try:
+        konfig_cls = konfigs[name]
+    except KeyError:
+        raise ValueError(f"Unable to find konfig with name{name!r}, candidates {list(konfigs.keys())}")
+
+    logger.info("Create instance of konfig %s", konfig_cls)
+    return konfig_cls(**extra)
+
+
+class secretstr(str):
+    def __repr__(self):
+        return f"<secretstr {'*' * len(self)}>"
+
+
+def fromenv(key: str, default=None, type: tp.Type = secretstr):
```

Comment:
I think that the prefix can be another argument with the default value == 'bf_'. It gives more flexibility and it's explicit. If you want to have a default prefix in your project you can always create a partial.

Reply:
And then you can delete the implicit mechanism

Reply:
I actually like this idea.
Maybe just completely remove this prefixing, user can always do it manually.
WDYT ?

Reply:
I think it's ok.

Reply:
(removing it)

Reply:
done